packages/text-field/Input.js

```diff
@@ -0,0 +1,195 @@
+    props.setBadInputHandler(this.badInputHandler);
+    props.setIsValidHandler(this.isValidHandler);
+  }
+
+  componentDidMount() {
+    this.props.handleValueChange(this.props.value);
+    if (this.props.id) {
+      this.props.setInputId(this.props.id);
+    }
+    if (this.props.disabled) {
+      this.props.setDisabled(true);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const {foundation} = nextProps;
+    this.handleValidationAttributeUpdate(nextProps);
+
+    if (this.props.disabled !== nextProps.disabled) {
+      nextProps.setDisabled(nextProps.disabled);
+      foundation.setDisabled(nextProps.disabled);
+    }
+
+    if (this.props.id !== nextProps.id) {
+      nextProps.setInputId(nextProps.id);
+    }
+  }
+
+  get classes() {
+    return classnames('mdc-text-field__input', this.props.className);
+  }
+
+  handleFocus = (e) => {
+    const {foundation, handleFocusChange, onFocus} = this.props;
+    foundation.activateFocus();
+    handleFocusChange(true);
+    onFocus(e);
+  }
+
+  handleBlur = (e) => {
+    const {foundation, handleFocusChange, onBlur} = this.props;
+    foundation.deactivateFocus();
+    handleFocusChange(false);
+    onBlur(e);
+  }
+
+  handleMouseDown = (e) => {
+    const {foundation, onMouseDown} = this.props;
+    foundation.setTransformOrigin(e);
+    onMouseDown(e);
+  }
+
+  handleTouchStart = (e) => {
+    const {foundation, onTouchStart} = this.props;
+    foundation.setTransformOrigin(e);
+    onTouchStart(e);
+  }
+
+  // To stay in sync with the MDC React Text Field Component, handleValueChange()
+  // is called to update MDC React Text Field's state. That state variable
+  // is used to let other subcomponents and the foundation know what the current
+  // value of the input is.
+  handleChange = (e) => {
+    const {foundation, handleValueChange, onChange} = this.props;
+    const {value} = e.target;
+
+    foundation.autoCompleteFocus();
+    handleValueChange(value);
+    onChange(e);
+  }
+
+  handleValidationAttributeUpdate = (nextProps) => {
+    const {foundation} = nextProps;
+    VALIDATION_ATTR_WHITELIST.some((attr) => {
+      if (this.props[attr] !== nextProps[attr]) {
+        // TODO: Update this when issue is fixed:
+        // https://github.com/material-components/material-components-web/issues/2716
+        foundation.handleValidationAttributeMutation_([{
+          // TODO: Update this when issue is fixed:
+          // https://github.com/material-components/material-components-web/issues/2717
+          attributeName: VALIDATION_ATTR_WHITELIST[0],
+        }]);
+        return true;
+      }
+    });
+  }
+
+  badInputHandler = () => this.inputElement.current.validity.badInput;
+  isValidHandler = () => this.inputElement.current.validity.valid;
+
+  render() {
+    const {
+      disabled,
+      /* eslint-disable no-unused-vars */
+      className,
+      foundation,
+      value,
+      handleFocusChange,
+      handleValueChange,
+      setBadInputHandler,
```

Comment:
Should this prop be 'handleBadInputChange'?

Reply:
I don't think so, because this method is passed a function that sets it on the Text Field. That function is what returns validity.badInput. We don't actually do any actions if badInput updates.